File: python/01_basic_chat_DEMETERICS.py
#!/usr/bin/env python3
"""
Example 1: Basic Chat with Demeterics Proxy

WHAT THIS DEMONSTRATES:
    - Using Demeterics observability proxy for AI calls
    - Automatic instrumentation without code changes
    - Analytics and monitoring for production AI
    - Transparent proxy pattern for any provider
    - Using Groq models through Demeterics

WHAT YOU'LL LEARN:
    - Routing API calls through observability layers
    - Production monitoring best practices
    - Analytics without modifying application code
    - Understanding proxy architectures
    - Getting insights into AI usage patterns

DEMETERICS PROVIDES:
    - Universal token usage tracking
    - Performance metrics (latency, queue times)
    - Cost analysis across all providers
    - Error tracking and debugging
    - Compliance audit trails
    - A/B testing capabilities
    - Usage analytics dashboard

PREREQUISITES:
    - Python 3.6 or higher
    - DEMETERICS_API_KEY environment variable set
    - Internet connection

EXPECTED OUTPUT:
    - Full JSON response (same as direct Groq)
    - AI's answer to your question
    - Token usage (tracked by Demeterics)
    - Visit Demeterics dashboard for analytics!

Run with: python3 01_basic_chat_DEMETERICS.py
"""

import http.client
import json
import os

# ==============================================================================
# Step 1: Load Demeterics API credentials
# ==============================================================================

api_key = os.environ.get('DEMETERICS_API_KEY')

if not api_key:
    print("Error: DEMETERICS_API_KEY environment variable not set")
    print("Run: export DEMETERICS_API_KEY='...'")
    print("Get your key from: https://demeterics.com")
    exit(1)

# Optional: Combined key format for advanced features
# If you have both Demeterics and provider keys:
# api_key = f"{demeterics_key};{groq_key}"

# ==============================================================================
# Step 2: Build the request payload (same as Groq)
# ==============================================================================

payload = {
    "model": "meta-llama/llama-4-scout-17b-16e-instruct",  # Groq model
    "messages": [
        {
            "role": "user",
            "content": "What is the capital of Switzerland?"
        }
    ],
    "temperature": 0.7,
    "max_tokens": 100
}

payload_json = json.dumps(payload)

# ==============================================================================
# Step 3: Connect to Demeterics proxy
# ==============================================================================

# Demeterics proxy endpoint (routes to Groq)
conn = http.client.HTTPSConnection("demeterics.uc.r.appspot.com")

# ==============================================================================
# Step 4: Send request through Demeterics
# ==============================================================================

headers = {
    'Content-Type': 'application/json',
    'Authorization': f'Bearer {api_key}'  # Demeterics API key
}

try:
    # Use Groq endpoint path, but through Demeterics proxy
    conn.request("POST", "/api/groq/v1/chat/completions", payload_json, headers)

    # Get the response
    response = conn.getresponse()
    response_data = response.read().decode('utf-8')

    # Parse JSON response
    response_json = json.loads(response_data)

    # ==============================================================================
    # Step 5: Display the results
    # ==============================================================================

    print("Full Response (via Demeterics proxy):")
    print(json.dumps(response_json, indent=2))

    # Extract and display the answer
    if 'choices' in response_json and len(response_json['choices']) > 0:
        answer = response_json['choices'][0]['message']['content']
        print("\n" + "="*50)
        print("AI's Answer (monitored by Demeterics):")
        print(answer)

    # Show token usage (tracked automatically)
    if 'usage' in response_json:
        usage = response_json['usage']
        print("\n" + "="*50)
        print("Token Usage (tracked):")
        print(f"  Prompt tokens: {usage.get('prompt_tokens', 0)}")
        print(f"  Completion tokens: {usage.get('completion_tokens', 0)}")
        print(f"  Total tokens: {usage.get('total_tokens', 0)}")
        print("\n📊 Analytics available at: https://demeterics.com/dashboard")

    # Check for Demeterics metadata
    if 'demeterics_metadata' in response_json:
        print("\n" + "="*50)
        print("Demeterics Metadata:")
        metadata = response_json['demeterics_metadata']
        print(f"  Request ID: {metadata.get('request_id')}")
        print(f"  Latency: {metadata.get('latency_ms')}ms")
        print(f"  Provider: {metadata.get('provider')}")

    # Check for errors
    if 'error' in response_json:
        print("\nAPI Error:")
        print(json.dumps(response_json['error'], indent=2))

except Exception as e:
    print(f"Request failed: {e}")
    print("\nTroubleshooting:")
    print("1. Check your internet connection")
    print("2. Verify your Demeterics API key")
    print("3. Check Demeterics service status")

finally:
    conn.close()

# ==============================================================================
# Demeterics Benefits:
# ==============================================================================
#
# Observability:
#   - Track every API call automatically
#   - Monitor token usage and costs
#   - Analyze performance metrics
#   - Debug production issues
#   - Generate compliance reports
#
# Architecture:
#   Your App → Demeterics Proxy → AI Provider
#                ↓
#          Analytics DB
#
# Features:
#   - Zero code changes required
#   - Works with existing integrations
#   - Universal format across providers
#   - Real-time dashboards
#   - Cost optimization insights
#   - Audit trails for compliance
#
# Use cases:
#   - Production monitoring
#   - Cost tracking and optimization
#   - Performance analysis
#   - A/B testing different models
#   - Compliance reporting
#   - Debugging and troubleshooting
#
# Dashboard features:
#   - Token usage by model/user/time
#   - Cost breakdown and projections
#   - Response time distributions
#   - Error rates and patterns
#   - Custom analytics queries
#

File: nodejs/01_basic_chat_DEMETERICS.js
/**
 * ============================================================================
 * Example 1: Basic Chat with Demeterics - Instrumented AI Proxy
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 * - How to use Demeterics observability proxy
 * - Automatic instrumentation of AI API calls
 * - Analytics and monitoring for production AI
 * - Using Groq models through Demeterics proxy
 *
 * WHAT YOU'LL LEARN:
 * - Routing API calls through an observability layer
 * - Getting analytics without code changes
 * - Understanding proxy patterns in AI deployments
 * - Production monitoring best practices
 *
 * DEMETERICS PROVIDES:
 * - Token usage tracking across all providers
 * - Performance metrics (latency, queue times)
 * - Cost analysis and optimization
 * - Error tracking and debugging
 * - Compliance audit trails
 * - A/B testing capabilities
 *
 * PREREQUISITES:
 * - Node.js installed (v14 or higher)
 * - DEMETERICS_API_KEY environment variable set
 *   Run: export DEMETERICS_API_KEY="..."
 *
 * EXPECTED OUTPUT:
 * - Full JSON response (same as direct Groq)
 * - The AI's answer to your question
 * - Token usage (tracked by Demeterics)
 * - Check Demeterics dashboard for analytics!
 *
 * RUN THIS:
 * node 01_basic_chat_DEMETERICS.js
 * ============================================================================
 */

// Step 1: Import the built-in HTTPS module
const https = require('https');

// Step 2: Prepare the request payload
// Same format as Groq (OpenAI-compatible)
const data = JSON.stringify({
  model: "meta-llama/llama-4-scout-17b-16e-instruct",  // Groq model via proxy
  messages: [
    {
      role: "user",
      content: "What is the capital of Switzerland?"
    }
  ],
  temperature: 0.7,   // Creativity level
  max_tokens: 100     // Maximum response length
});

// Step 3: Configure the Demeterics proxy request
// Note: Different hostname, same API path as Groq
const options = {
  hostname: 'demeterics.uc.r.appspot.com',     // Demeterics proxy server
  port: 443,
  path: '/api/groq/v1/chat/completions',       // Groq endpoint via proxy
  method: 'POST',
  headers: {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${process.env.DEMETERICS_API_KEY}`,  // Demeterics key
    'Content-Length': Buffer.byteLength(data)
  }
};

// Optional: If you have both Demeterics AND Groq keys, you can use combined format:
// 'Authorization': `Bearer ${process.env.DEMETERICS_API_KEY};${process.env.GROQ_API_KEY}`
// This enables advanced features like fallback and provider switching

// Step 4: Create and send the request
const req = https.request(options, (res) => {
  let responseData = '';

  // Accumulate response chunks
  res.on('data', (chunk) => {
    responseData += chunk;
  });

  // Process complete response
  res.on('end', () => {
    try {
      const parsed = JSON.parse(responseData);

      // Display full response
      console.log('Full Response (via Demeterics):');
      console.log(JSON.stringify(parsed, null, 2));

      // Extract the AI's answer
      if (parsed.choices && parsed.choices[0]) {
        console.log('\nAI Answer:');
        console.log(parsed.choices[0].message.content);
      }

      // Show token usage (tracked by Demeterics)
      if (parsed.usage) {
        console.log('\nToken Usage (monitored):');
        console.log(`Prompt: ${parsed.usage.prompt_tokens}`);
        console.log(`Completion: ${parsed.usage.completion_tokens}`);
        console.log(`Total: ${parsed.usage.total_tokens}`);
        console.log('\n📊 Check Demeterics dashboard for detailed analytics!');
      }

      // Demeterics may add metadata
      if (parsed.demeterics_metadata) {
        console.log('\nDemeterics Metadata:');
        console.log(JSON.stringify(parsed.demeterics_metadata, null, 2));
      }

    } catch (error) {
      console.error('Error parsing response:', error.message);
      console.log('Raw response:', responseData);
    }
  });
});

// Handle request errors
req.on('error', (error) => {
  console.error('Request failed:', error.message);

  if (error.message.includes('ENOTFOUND')) {
    console.log('Check your internet connection');
  }
  if (!process.env.DEMETERICS_API_KEY) {
    console.log('Make sure DEMETERICS_API_KEY is set');
  }
});

// Send the request
req.write(data);
req.end();

// Demeterics Benefits:
// - Zero code changes for observability
// - Works with existing API integrations
// - Universal analytics across all providers
// - Production debugging and monitoring
// - Cost optimization insights
// - Compliance and audit trails
//
// Architecture:
// Your App → Demeterics Proxy → AI Provider (Groq/OpenAI/etc)
//              ↓
//         Analytics DB
//
// The proxy is transparent - same API format, same responses
// But now with complete observability!